staging/src/k8s.io/client-go/util/watchlist/watch_list.go

```diff
@@ -0,0 +1,77 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package watchlist
+
+import (
+	metainternalversion "k8s.io/apimachinery/pkg/apis/meta/internalversion"
+	metainternalversionvalidation "k8s.io/apimachinery/pkg/apis/meta/internalversion/validation"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	clientfeatures "k8s.io/client-go/features"
+	"k8s.io/utils/ptr"
+)
+
+// CanUseWatchListForListRequest determines if the given
+// ListOptions can be used to form a watch-list request,
+// which would result in streaming semantically equal data from the server.
+func CanUseWatchListForListRequest(listOptions metav1.ListOptions, scheme *runtime.Scheme) (bool, error) {
+	if !clientfeatures.FeatureGates().Enabled(clientfeatures.WatchListClient) {
+		return false, nil
+	}
+
+	// to ensure that we can create a watch-list request that returns
+	// semantically equivalent data for the given listOptions,
+	// we need to validate that the RVM for the list is supported by watch-list requests.
+	if listOptions.ResourceVersionMatch != "" && listOptions.ResourceVersionMatch != metav1.ResourceVersionMatchNotOlderThan {
+		return false, nil
+	}
+	if listOptions.Limit > 0 {
+		return false, nil
+	}
+
+	internalListOptions := &metainternalversion.ListOptions{}
+	if err := scheme.Convert(&listOptions, internalListOptions, nil); err != nil {
+		return false, err
+	}
+	if errs := metainternalversionvalidation.ValidateListOptions(internalListOptions, true); len(errs) > 0 {
+		return false, nil
+	}
+
+	internalWatchListOptions := &metainternalversion.ListOptions{}
+	watchListOptions := PrepareWatchListOptionsFromListOptions(listOptions)
+	if err := scheme.Convert(&watchListOptions, internalWatchListOptions, nil); err != nil {
```

Comment:
question: would it work if we just directly call: `Convert_v1_ListOptions_To_internal_ListOptions` instead?

If that works - we wouldn't even have to register anything more in schema and pass the schema itself.